### PR TITLE
Distribute staking rewards

### PR DESCRIPTION
**Pull request by pro-wh**
It was stated in the incentive design that we would reward accounts for having escrowed stake.

```
hook onEpochChange():
  for a in accounts:
    a.escrowed += amountEligibleStake(a) * rewardRate
  butLessOverTime()
```

blocks #2265 
depends #2303

### CODE REVIEW

**Comment by pro-wh** [comment]
In this version:

- reward tokens come out of thin air
- no events emitted
- we modify the state tree values while we iterate over a range--is that legal?

**Comment by pro-wh** [comment]
I changed this to pay rewards from the common pool instead of from thin air

File: go/staking/api/api.go [hunk 1]
```diff
@@ -3,7 +3,8 @@ package api
 
 import (
 	"context"
-	"errors"
+
+	"github.com/pkg/errors"
```
Comment (kostko) [inline go/staking/api/api.go]: Why this change?

File: go/tendermint/apps/staking/state/state.go [hunk 2]
```diff
@@ -62,10 +62,14 @@ var (
 	//
 	// Value is empty.
 	debondingQueueKeyFmt = keyformat.New(0x57, uint64(0), &signature.MapKey{}, &signature.MapKey{}, uint64(0))
+	// rewardScheduleKeyFmt is the key format used for the reward schedule.
+	//
+	// Value is a CBOR-serialized slice of staking.RewardStep
```
Comment (kostko) [inline go/tendermint/apps/staking/state/state.go]:
```suggestion
	// Value is a CBOR-serialized slice of staking.RewardStep.
```

File: go/tendermint/apps/staking/state/state.go [hunk 3]
```diff
@@ -525,6 +547,82 @@ func (s *MutableState) TransferFromCommon(ctx *abci.Context, toID signature.Publ
...
+	if stepRelativeTime%activeStep.Interval != 0 {
+		return nil
+	}
+
+	commonPool, err := s.CommonPool()
+	if err != nil {
+		return errors.Wrap(err, "loading common pool")
+	}
+
+	s.Snapshot.IterateRange(accountKeyFmt.Encode(), nil, true, func(key, value []byte) bool {
```
Comment (kostko) [inline go/tendermint/apps/staking/state/state.go]: This seems to go over all accounts? Shouldn't we only reward accounts that belong to entities running validator nodes? Otherwise people can decide to delegate to escrow accounts that don't run any nodes and thus get all the rewards without taking on any risk (as those accounts will never be slashed).

File: go/tendermint/apps/staking/state/state.go [hunk 4]
```diff
@@ -525,6 +547,82 @@ func (s *MutableState) TransferFromCommon(ctx *abci.Context, toID signature.Publ
...
+			err = errors.Wrap(err1, "dividing by reward step denominator")
+			return true
+		}
+
+		if q.IsZero() {
+			return false
+		}
+
+		if err1 := staking.Move(&ent.Escrow.Active.Balance, commonPool, q); err1 != nil {
+			err = errors.Wrap(err1, "transferring to active escrow balance from common pool")
```
Comment (kostko) [inline go/tendermint/apps/staking/state/state.go]: So if there is not enough funds to pay out the rewards, this will cause a consensus halt?

File: go/tendermint/apps/staking/state/state.go (hunk 3 above)
Comment (kostko) [reply go/tendermint/apps/staking/state/state.go]: Maybe we should record all nodes that were active for larger part of an epoch and only distribute rewards to accounts belonging to their entities.

File: go/staking/api/api.go (hunk 1 above)
Comment (pro-wh) [reply go/staking/api/api.go]: oh I was going to add some context to some errors, but then I refactored that stuff into other files. I'll revert this for now

File: go/tendermint/apps/staking/state/state.go (hunk 3 above)
Comment (pro-wh) [reply go/tendermint/apps/staking/state/state.go]: ok I'll work on that

File: go/tendermint/apps/staking/state/state.go (hunk 2 above)
Comment (pro-wh) [reply go/tendermint/apps/staking/state/state.go]: gonna include this in the next push

**Comment by pro-wh** [comment]
in this version:
- rewards distributed as part of election
  - it's nice to have it all there so that you can't evade elections while getting rewards
  - if a node of yours was considered eligible for any committee or validator, then you're eligible
- no more separate reward interval
- reward step now has a scaling factor that affects rewards distributed in that interval
- moved denominator to a shared variable 🤷

File: go/tendermint/apps/staking/state/state.go (hunk 3 above)
Comment (pro-wh) [reply go/tendermint/apps/staking/state/state.go]: went with distributing reward when an entity's node is considered for election

File: go/staking/api/rewards.go [hunk 5]
```diff
@@ -0,0 +1,16 @@
+package api
+
+import (
+	"math/big"
+)
+
+var RewardAmountDenominator *Quantity
```
Comment (kostko) [inline go/staking/api/rewards.go]: Add documentation block.

File: go/tendermint/apps/scheduler/scheduler.go [hunk 6]
```diff
@@ -40,6 +41,8 @@ var (
 	rngContextMerge                = []byte("EkS-ABCI-Merge")
 	rngContextValidators           = []byte("EkS-ABCI-Validators")
 
+	RewardFactorEpochElectionAny *staking.Quantity
```
Comment (kostko) [inline go/tendermint/apps/scheduler/scheduler.go]: Add documentation block. Is this something that should be part of the API or is it specific to the tendermint implementation?

File: go/tendermint/apps/staking/state/state.go [hunk 7]
```diff
@@ -543,6 +565,64 @@ func (s *MutableState) TransferFromCommon(ctx *abci.Context, toID signature.Publ
...
+
+		if q.IsZero() {
+			continue
+		}
+
+		if err := staking.Move(&ent.Escrow.Active.Balance, commonPool, q); err != nil {
+			return errors.Wrap(err, "transferring to active escrow balance from common pool")
+		}
+
+		s.SetAccount(id, ent)
```
Comment (kostko) [inline go/tendermint/apps/staking/state/state.go]: We should do all updates at the end when we are sure that there are no errors (or use a snapshot?). Otherwise this can cause stuff to still be committed (e.g., there is no implicit rollback in case of errors).

Note that in the current implementation this is probably not a problem as any error here will cause a consensus failure and nothing will be committed. But in case this is ever used outside of `BeginBlock` it may be useful to mitigate.

Another reason for deferring updates is non-determinism -- IAVL is sensitive to order of updates, so we (currently) need to sort the updates before committing them. _(An alternative is to have an intermediate layer that caches and sorts all IAVL updates before they are committed or to use our own data structure, e.g. #1898, which doesn't have this problem. But this is out of scope for this PR.)_

File: go/tendermint/apps/staking/state/state.go [hunk 8]
```diff
@@ -543,6 +565,64 @@ func (s *MutableState) TransferFromCommon(ctx *abci.Context, toID signature.Publ
...
+		// We're past the end of the schedule.
+		return nil
+	}
+
+	commonPool, err := s.CommonPool()
+	if err != nil {
+		return errors.Wrap(err, "loading common pool")
+	}
+
+	for mk := range accounts {
```
Comment (kostko) [inline go/tendermint/apps/staking/state/state.go]: Iteration over Go maps is non-deterministic and may cause consensus failures (different validators will produce different hashes due to IAVL being sensitive to order of updates). One idea is to add updates to a list and sort them before committing -- this would also solve the issue described below about possibly committing partial updates.

File: go/tendermint/apps/staking/state/state.go (hunk 7 above)
Comment (pro-wh) [reply go/tendermint/apps/staking/state/state.go]: is there a place outside of BeginBlock that would disrespect an error condition and proceed to act on the same tree? I'd consider continuing in that condition to be the problem.

I'll treat the nondeterministic map iteration as a separate issue and discuss that in your other comment.

File: go/tendermint/apps/staking/state/state.go (hunk 8 above)
Comment (pro-wh) [reply go/tendermint/apps/staking/state/state.go]: I'll sort these

File: go/tendermint/apps/scheduler/scheduler.go (hunk 6 above)
Comment (pro-wh) [reply go/tendermint/apps/scheduler/scheduler.go]: ah yeah this would be better in API land where we've been keeping the denominators

File: go/staking/api/rewards.go (hunk 5 above)
Comment (pro-wh) [reply go/staking/api/rewards.go]: ok

File: go/tendermint/apps/scheduler/scheduler.go (hunk 6 above)
Comment (pro-wh) [reply go/tendermint/apps/scheduler/scheduler.go]: or, more consistent, at least

File: go/tendermint/apps/staking/state/state.go (hunk 7 above)
Comment (kostko) [reply go/tendermint/apps/staking/state/state.go]: > is there a place outside of BeginBlock that would disrespect an error condition and proceed to act on the same tree? I'd consider continuing in that condition to be the problem.

Potentially in DeliverTx. An error there aborts the transaction but should not commit any state changes.

We could enforce that at the mux layer, but then we need a way to still have some state side effects even for failing transactions (because maybe sometimes you would want some even then).

Comment (pro-wh) [reply go/tendermint/apps/staking/state/state.go]: > aborts the transaction but should not commit any state changes

aw ew

probably better to snapshot, so that we have slightly less "have all accounts in memory" situations, in case we ever have to get rid of them

Comment (kostko) [reply go/tendermint/apps/staking/state/state.go]: Yeah makes sense. So for now I see that you just have a comment that says that accounts may be corrupted in case of errors. Maybe just add something like "if this is an issue, the caller should snapshot and rollback in case of errors".

Comment (pro-wh) [reply go/tendermint/apps/staking/state/state.go]: added

**Comment by pro-wh** [comment]
Gonna merge. This will not set a reward schedule in any testing or production genesis document, so it will not actually distribute any rewards on its own.
